### PR TITLE
snap: update security-secrets-setup config

### PR DESCRIPTION
**Pull request by tonyespy**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #2547 

Also see related issue: #2495 

## What is the new behavior?
The new behavior is that PKI setup (aka security-secrets-setup) in the snap is now aligned with the docker deployment of Geneva, and the intermediate files in `$SNAP_DATA/pki/EdgeXFoundryCA` are no longer present.  See the issue for more details.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Testing Instructions

- [ ] Test that the snap builds and installs successfully
- [ ] Test that the expected certificates are generated inside $SNAP_DATA/secrets

```
$ export SNAP_DATA=/var/snap/edgexfoundry/current
espy@sumo$ sudo tree $SNAP_DATA/secrets/ -P "*.pem|*.crt|*.key" --prune
/var/snap/edgexfoundry/current/secrets/
├── ca
│   └── ca.pem
├── edgex-kong
│   ├── ca.pem
│   ├── server.crt
│   └── server.key
└── edgex-vault
    ├── ca.pem
    ├── server.crt
    └── server.key

3 directories, 7 files
```

Specifically, **ca.pem** (the root CA) and the TLS certificates (**server.crt**) and keys (**server.key**) for kong and vault.

- [x] Test that one can use the certificate from `$SNAP_DATA/secrets/ca/ca.pem` to connect with TLS host verification on with Vault and Kong

verify the curl generates a self-signed certificate error:
```
$ curl https://localhost:8443/
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

copy the CA to a user-readable area:
```
$ sudo cat /var/snap/edgexfoundry/current/secrets/ca/ca.pem > ca.pem
```

Kong uses a cert signed by the CA (on port 8443):
```
$ curl --cacert ca.pem https://localhost:8443; echo
{"message":"no Route matched with those values"}
``` 
Kong uses the specific cert from the edgex-kong folder (note you need `-servername` here to set the Host header and thus have Kong respond with the right cert, it can use different certs for different Host/SNIs settings):
```
$ sudo cat /var/snap/edgexfoundry/current/secrets/edgex-kong/server.crt > expected-kong.crt
$ openssl s_client -connect localhost:8443 -servername localhost 2>/dev/null </dev/null |  sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > kong.crt
$ diff -u kong.crt expected-kong.crt
```

Vault uses a cert signed by the CA (on port 8200):
```
$ curl --cacert ca.pem https://localhost:8200; echo 
<a href="/ui/">Temporary Redirect</a>.


```

Vault uses the specific cert from edgex-vault folder:
```
$ openssl s_client -connect localhost:8200 -servername localhost 2>/dev/null </dev/null |  sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > vault.pem
$ sudo cat /var/snap/edgexfoundry/current/secrets/edgex-vault/server.crt > expected-vault.crt
$ diff -u vault.crt expected-vault.crt
```

- [ ] Test using an authentication token to access a core-* REST API thru Kong

```
$ curl --cacert ca.pem https://localhost:8443/command/api/v1/ping
{"message":"Unauthorized"}$
```
generate an auth token:

```
$ sudo edgexfoundry.security-proxy-setup-cmd --confdir $SNAP_DATA/config/security-proxy-setup/res/ --useradd=me --group=admin | grep "access token"
the access token for user me is: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ2Mmo3R3dzME1sd2ZnQ1F0bnhTMDR5N2NtSG5QZlgwVyIsImFjY291bnQiOiJtT1VaRVkifQ.4SjMGPYDMYF1KZEcuDyJtp-8G-O6bSd3030RyIxGNlU. Please keep the token for accessing edgex services
```
Note - your token will be different.

```
$ TOKEN=<cut & paste token from previous step>
$ curl --cacert ca.pem --header "Authorization: Bearer $TOKEN" \
https://localhost:8443/command/api/v1/ping
pong$
```

### CODE REVIEW

**Comment by anonymouse64** [comment]
Is there a reason why we don't make `/var/snap/edgexfoundry/current/secrets/ca/ca.pem`  world-readable? I've always been slightly annoyed by having to copy the cert somewhere to use it. Also, not for this PR, but we should probably make the CA readable by other snaps via the content interface.

**Comment by bnevis-i** [comment]
> Is there a reason why we don't make `/var/snap/edgexfoundry/current/secrets/ca/ca.pem` world-readable? I've always been slightly annoyed by having to copy the cert somewhere to use it. Also, not for this PR, but we should probably make the CA readable by other snaps via the content interface.

 umask issue?  In Docker, the CA ends up world readable; same with the certificates.

```
ls -lR
.:
total 12
drwxr-xr-x 2 root            root   4096 May 10 16:07 ca
drwxr-xr-x 2 root            root   4096 May 10 16:07 edgex-kong
drwxr-xr-x 2 systemd-network bnevis 4096 May 10 16:07 edgex-vault

./ca:
total 4
-rw-r--r-- 1 root root 960 May 10 16:07 ca.pem

./edgex-kong:
total 12
-rw-r--r-- 1 root root 960 May 10 16:07 ca.pem
-rw-r--r-- 1 root root 993 May 10 16:07 server.crt
-r-------- 1 root root 306 May 10 16:07 server.key

./edgex-vault:
total 12
-rw-r--r-- 1 systemd-network bnevis  960 May 10 16:07 ca.pem
-rw-r--r-- 1 systemd-network bnevis 1001 May 10 16:07 server.crt
-r-------- 1 systemd-network bnevis  306 May 10 16:07 server.key
```

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -127,7 +127,7 @@ apps:
       KONG_ADMIN_LISTEN: "0.0.0.0:8001, 0.0.0.0:8444 ssl"
       LC_ALL: C.UTF-8
       LANG: C.UTF-8
-    start-timeout: 15m
+    start-timeout: 7m
```
Comment (anonymouse64) [inline snap/snapcraft.yaml:130]: why is this changed here? I don't see anything in the commit message about why this was changed?

Comment (tonyespy) [reply snap/snapcraft.yaml:130]: Nice catch, this should've been cleaned up. I was getting tired of waiting 15 minutes for service failures when debugging my changes. FYI, I have a separate trello card to investigate whether we can finally crank those done to something more reasonable and/or make them configurable.

Comment (tonyespy) [reply snap/snapcraft.yaml]: OK, just re-pushed and changed these all back to 15m.

**Comment by tonyespy** [comment]
Also just a heads-up, based upon my testing over this past weekend, I've come to realize that deploying EdgeX in production environments is a bit tricky due to the current PKI/vault/kong implementation. As such, I added commentary to issue #1926 (enable end-user Kong TLS certs), stating that I'd like to see this prioritized for Hanoi.

**Comment by tonyespy** [comment]
> Is there a reason why we don't make `/var/snap/edgexfoundry/current/secrets/ca/ca.pem` world-readable? I've always been slightly annoyed by having to copy the cert somewhere to use it. Also, not for this PR, but we should probably make the CA readable by other snaps via the content interface.

It's already world-readable in the snap:
```

espy@sumo$ pwd
/var/snap/edgexfoundry/current/secrets/ca
espy@sumo$ ls -la
total 16
drwxr-xr-x  2 root root 4096 May 11 10:52 .
drwxr-xr-x 20 root root 4096 May 11 10:53 ..
-rw-r--r--  1 root root  956 May 11 10:52 ca.pem
-r--------  1 root root   10 May 11 10:52 .security-secrets-setup.complete

```

**Comment by anonymouse64** [comment]
@tonyespy 
> It's already world-readable in the snap:

I was just going off your testing instructions from the PR description, which say:

> copy the CA to a user-readable area:
> ```
> $ sudo cat /var/snap/edgexfoundry/current/secrets/ca/ca.pem > ca.pem
> ```

**Comment by tonyespy** [comment]
> @tonyespy
> 
> > It's already world-readable in the snap:
> 
> I was just going off your testing instructions from the PR description, which say:
> 
> > copy the CA to a user-readable area:
> > ```
> > $ sudo cat /var/snap/edgexfoundry/current/secrets/ca/ca.pem > ca.pem
> > ```

And I just copied those instructions from your previous testing instructions in old PR, clearly not required anymore! ;)-

**Comment by tonyespy** [comment]
recheck

**Comment by bnevis-i** [comment]
> recheck

I don't think recheck works any longer.  Ammend your commit and force push it and I will reapprove.
